app/src/main/java/pl/futurecollars/invoicing/db/file/FileBasedInvoiceRepository.java

```diff
@@ -0,0 +1,140 @@
+package pl.futurecollars.invoicing.db.file;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import java.io.IOException;
+import java.nio.file.StandardOpenOption;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import pl.futurecollars.invoicing.db.InvoiceRepository;
+import pl.futurecollars.invoicing.exceptions.FileSystemException;
+import pl.futurecollars.invoicing.model.Invoice;
+import pl.futurecollars.invoicing.services.FileService;
+import pl.futurecollars.invoicing.services.JsonService;
+
+public class FileBasedInvoiceRepository implements InvoiceRepository {
+
+    private final FileService fileService;
+    private final JsonService<Invoice> jsonService;
+
+    public FileBasedInvoiceRepository(FileService fileService, JsonService<Invoice> jsonService) {
+        this.fileService = fileService;
+        this.jsonService = jsonService;
+    }
+
+    @Override
+    public Invoice save(Invoice invoice) {
+        UUID id = UUID.randomUUID();
+        try {
+            if (!this.getById(id).isPresent()) {
+                invoice.setId(id);
+                String json = jsonService.toJson(invoice);
+                fileService.writeToFile(json);
+                return invoice;
+            } else {
+                return save(invoice);
+            }
+        } catch (IOException e) {
+            throw new FileSystemException("File system processing error.");
+        }
+    }
+
+    @Override
+    public Optional<Invoice> getById(UUID id) {
+        List<Invoice> invoices;
+        try {
+            invoices = fileService.readFile()
+                .map(item -> {
+                    try {
+                        return jsonService.toObject(item, Invoice.class);
+                    } catch (JsonProcessingException e) {
+                        throw new FileSystemException("Json processing error.");
```

Comment:
zamiast tak paskudzić lambdy tymi trycatchami - zrób tego try catch w jsonService, ewentualnie zamien go na Unchecked exception zamiast aktualnego checked exception.

Reply:
Checked exception powinno sie wyłapać tak szybko jak to tylko możliwe i zmienić na unchecked, aby własnie unikać takich przekombinowanych lambd